File: go/test/user_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/mock-api-service-sdk/go"
	"github.com/voxgig-sdk/mock-api-service-sdk/go/core"

	vs "github.com/voxgig-sdk/mock-api-service-sdk/go/utility/struct"
)

func TestUserEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.User(nil)
		if ent == nil {
			t.Fatal("expected non-nil UserEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := userBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "user." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set MOCKAPISERVICE_TEST_USER_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		userRef01Ent := client.User(nil)
		userRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "user"}, setup.data), "user_ref01"))

		userRef01DataResult, err := userRef01Ent.Create(userRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		userRef01Data = core.ToMapAny(userRef01DataResult)
		if userRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if userRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		userRef01Match := map[string]any{}

		userRef01ListResult, err := userRef01Ent.List(userRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		userRef01List, userRef01ListOk := userRef01ListResult.([]any)
		if !userRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", userRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(userRef01List), map[string]any{"id": userRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		userRef01DataUp0Up := map[string]any{
			"id": userRef01Data["id"],
		}

		userRef01MarkdefUp0Name := "created_at"
		userRef01MarkdefUp0Value := fmt.Sprintf("Mark01-user_ref01_%d", setup.now)
		userRef01DataUp0Up[userRef01MarkdefUp0Name] = userRef01MarkdefUp0Value

		userRef01ResdataUp0Result, err := userRef01Ent.Update(userRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		userRef01ResdataUp0 := core.ToMapAny(userRef01ResdataUp0Result)
		if userRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if userRef01ResdataUp0["id"] != userRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if userRef01ResdataUp0[userRef01MarkdefUp0Name] != userRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", userRef01MarkdefUp0Name, userRef01ResdataUp0[userRef01MarkdefUp0Name])
		}

		// LOAD
		userRef01MatchDt0 := map[string]any{
			"id": userRef01Data["id"],
		}
		userRef01DataDt0Loaded, err := userRef01Ent.Load(userRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		userRef01DataDt0LoadResult := core.ToMapAny(userRef01DataDt0Loaded)
		if userRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if userRef01DataDt0LoadResult["id"] != userRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		userRef01MatchRm0 := map[string]any{
			"id": userRef01Data["id"],
		}
		_, err = userRef01Ent.Remove(userRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		userRef01MatchRt0 := map[string]any{}

		userRef01ListRt0Result, err := userRef01Ent.List(userRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		userRef01ListRt0, userRef01ListRt0Ok := userRef01ListRt0Result.([]any)
		if !userRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", userRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(userRef01ListRt0), map[string]any{"id": userRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func userBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "user", "UserTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read user test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse user test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"user01", "user02", "user03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("MOCKAPISERVICE_TEST_USER_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"MOCKAPISERVICE_TEST_USER_ENTID": idmap,
		"MOCKAPISERVICE_TEST_LIVE":      "FALSE",
		"MOCKAPISERVICE_TEST_EXPLAIN":   "FALSE",
		"MOCKAPISERVICE_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["MOCKAPISERVICE_TEST_USER_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["MOCKAPISERVICE_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["MOCKAPISERVICE_APIKEY"],
			},
			extra,
		})
		client = sdk.NewMockApiServiceSDK(core.ToMapAny(mergedOpts))
	}

	live := env["MOCKAPISERVICE_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["MOCKAPISERVICE_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/user_entity_test.lua
-- User entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("mock-api-service_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("UserEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:User(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = user_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "user." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set MOCKAPISERVICE_TEST_USER_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local user_ref01_ent = client:User(nil)
    local user_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.user"), "user_ref01"))

    local user_ref01_data_result, err = user_ref01_ent:create(user_ref01_data, nil)
    assert.is_nil(err)
    user_ref01_data = helpers.to_map(user_ref01_data_result)
    assert.is_not_nil(user_ref01_data)
    assert.is_not_nil(user_ref01_data["id"])

    -- LIST
    local user_ref01_match = {}

    local user_ref01_list_result, err = user_ref01_ent:list(user_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(user_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(user_ref01_list_result),
      { id = user_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local user_ref01_data_up0_up = {
      id = user_ref01_data["id"],
    }

    local user_ref01_markdef_up0_name = "created_at"
    local user_ref01_markdef_up0_value = "Mark01-user_ref01_" .. tostring(setup.now)
    user_ref01_data_up0_up[user_ref01_markdef_up0_name] = user_ref01_markdef_up0_value

    local user_ref01_resdata_up0_result, err = user_ref01_ent:update(user_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local user_ref01_resdata_up0 = helpers.to_map(user_ref01_resdata_up0_result)
    assert.is_not_nil(user_ref01_resdata_up0)
    assert.are.equal(user_ref01_resdata_up0["id"], user_ref01_data_up0_up["id"])
    assert.are.equal(user_ref01_resdata_up0[user_ref01_markdef_up0_name], user_ref01_markdef_up0_value)

    -- LOAD
    local user_ref01_match_dt0 = {
      id = user_ref01_data["id"],
    }
    local user_ref01_data_dt0_loaded, err = user_ref01_ent:load(user_ref01_match_dt0, nil)
    assert.is_nil(err)
    local user_ref01_data_dt0_load_result = helpers.to_map(user_ref01_data_dt0_loaded)
    assert.is_not_nil(user_ref01_data_dt0_load_result)
    assert.are.equal(user_ref01_data_dt0_load_result["id"], user_ref01_data["id"])

    -- REMOVE
    local user_ref01_match_rm0 = {
      id = user_ref01_data["id"],
    }
    local _, err = user_ref01_ent:remove(user_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local user_ref01_match_rt0 = {}

    local user_ref01_list_rt0_result, err = user_ref01_ent:list(user_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(user_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(user_ref01_list_rt0_result),
      { id = user_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function user_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/user/UserTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read user test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "user01", "user02", "user03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("MOCKAPISERVICE_TEST_USER_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["MOCKAPISERVICE_TEST_USER_ENTID"] = idmap,
    ["MOCKAPISERVICE_TEST_LIVE"] = "FALSE",
    ["MOCKAPISERVICE_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["MOCKAPISERVICE_TEST_USER_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["MOCKAPISERVICE_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["MOCKAPISERVICE_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["MOCKAPISERVICE_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
